### PR TITLE
[Inductor] Support vectorized transpose in CPP backend

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91532

Fix https://github.com/pytorch/torchdynamo/issues/1915
This PR adds the vectorization support for transposed operations in TorchInductor CPP backend. It contains the following changes:
1. `CppTile2DKernelChecker` is added to check the eligibility of applying the optimization. We only addresss a narrow set of situations. All of the following conditions should be met: 1) There exists one and only one fp32 load/store with outer loop var having contiguous buffer accesses. 2) When a load/store doesn't have contiguous access in an outer loop var, the access should be vectorizable from the inner-most dim. 3) No reduction. More scenarios/operations would be supported in the future PRs.
2. If `CppTile2DKernelChecker` reports the optimization is doable, `CppKernelProxy` would split/tile the loops from both the outer loop var having contiguous buffer access and the inner-most loop var.
3. The main loop split from the outer loop var is further split at the inner-most level and then handled by `CppTile2DKernel` and `CppTile2DTailKernel` which generate the transposed load/store. The former kernel does the vectorized transposed load/store on tiles and then does vectorized load/store/compute along the inner-most loop axis. The vectorized transpose micro-kernel implementation borrows/refers to that from FBGEMM. The latter kernel simply does scalar operations.
4. The tail loop split from the outer loop var directly calls `CppKernel` with scalar operations.

Next steps:
1. Support vectorized transpose with smaller tile size at one dim but bigger tile size at the other, e.g., 3x784.
2. Support reduction vectorized on the outer loop var (contiguous from outer loop var, not with inner-most loop var)

cc @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by EikanWang — APPROVED** [review]

File: torch/_inductor/codegen/cpp_prefix.h [hunk 1]
```diff
@@ -69,3 +70,340 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
     dst_u32[i] = *(src + i) ? 0xFFFFFFFF : 0;
   }
 }
+
+#if defined(CPU_CAPABILITY_AVX512)
+// TODO(jgong5): rewrite with ATEN vectorized (need to add unpack and shuffle)
+// Code from FBGEMM:
+// https://github.com/pytorch/FBGEMM/blob/39a423e4ad1a04b77fea81c7d09c3e6f8984fae9/src/UtilsAvx512.cc#LL19C6-L19C6
+// 16 * 6 = 96 instructions
+inline void transpose_kernel_16x16_avx512(
```
Comment (jansel) [inline torch/_inductor/codegen/cpp_prefix.h:79]: It feels like this should be part of `at::vec` as opposed to in this header.

File: torch/_inductor/codegen/cpp.py [hunk 2]
```diff
@@ -1049,45 +1224,187 @@ def to_dtype(x, dtype):
         return self
 
 
+class CppTile2DKernelChecker(CppVecKernelChecker):
+    """
+    We only addresss a narrow set of situations. All of the following conditions should be met:
+    1. There are only loads and stores.
```
Comment (jansel) [inline torch/_inductor/codegen/cpp.py:1230]: How will this extend to kernels that do compute?  Can we do this in a more general way?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -902,9 +911,176 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
...
+        else:
+            ld_dst = f"{cexpr(self.stride_at(self.itervars[non_contig_idx], expanded_index))}"
+            ld_src = f"{self.tiling_factor}"
+            line = f"TILE2D_COPY_TRANSPOSE({var} + {cexpr(new_index)}, {value}, {ld_dst}, {ld_src});"
+
+        self.stores.writeline(name, line)
+
+
+@load_store_only_kernel
+class CppTile2DTailKernel(CppKernel):
```
Comment (jansel) [inline torch/_inductor/codegen/cpp.py:1043]: Add docstrings

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -902,9 +911,176 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
...
+        return self
+
+    cls.__exit__ = __exit__
+    cls.__enter__ = __enter__
+
+    return cls
+
+
+@load_store_only_kernel
+class CppTile2DKernel(CppVecKernel):
```
Comment (jansel) [inline torch/_inductor/codegen/cpp.py:947]: Add docstrings

File: torch/_inductor/codegen/cpp_prefix.h [hunk 5]
```diff
@@ -69,3 +70,340 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
...
+
+#endif
+
+// copy tile_size*tile_size data from src to dst
+#define TILE2D_COPY(dst, src, offset, tile_size)                                           \
+  for (long offset = 0; offset < tile_size; offset++) {                                    \
+    auto tmp = at::vec::Vectorized<                                                        \
+      typename std::remove_extent<typename std::remove_pointer<decltype(src)>::type>::type \
+    >::loadu(src);                                                                         \
+    tmp.store(dst);                                                                        \
```
Comment (jansel) [inline torch/_inductor/codegen/cpp_prefix.h:408]: Baking the load and store into a single macro will make this hard to generalize to kernels that have compute.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp_prefix.h (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp_prefix.h]: Moved to `at::vec`.

File: torch/_inductor/codegen/cpp.py (hunk 2 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py]: Generalize the implementation and added compute support too.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:1060]: Added.

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:946]: Added.

File: torch/_inductor/codegen/cpp_prefix.h (hunk 5 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp_prefix.h]: Generalize the implementation. Now the vectorized load/store/compute reuses the logic from `CppVecKernel`.

**Comment by jgong5** [comment]
CI failures seem irrelevant to the PR changes.

File: torch/_inductor/codegen/cpp_prefix.h [hunk 6]
```diff
@@ -69,3 +69,19 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
     dst_u32[i] = *(src + i) ? 0xFFFFFFFF : 0;
   }
 }
+
+#if defined(CPU_CAPABILITY_AVX512)
+
+#define TILE2D_COPY_TRANSPOSE(dst, src, ld_dst, ld_src)       \
```
Comment (jansel) [inline torch/_inductor/codegen/cpp_prefix.h:75]: Why do we need these macros?  Perhaps call at::vec directly?

Comment (jgong5) [reply torch/_inductor/codegen/cpp_prefix.h]: Thanks for the comment. I intended to let the generated code invoke something with general name. Now I reverted this and added the general `transpose_mxn` template function into at::vec. Please review. Thanks.

**Comment by jgong5** [comment]
CI failures are irrelevant.

**Review by jansel — APPROVED** [review]
Approving with understand this will be cleaned up in a future PR

**Comment by jgong5** [comment]
> Approving with understand this will be cleaned up in a future PR

Thanks for all the comments/suggestions on the PR and offline. I will follow up with the tile abstraction for a cleaner design in the follow-up PRs.

**Pull request closed without merging**